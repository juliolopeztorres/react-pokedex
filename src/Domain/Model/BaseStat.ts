
export default class BaseStat {

  constructor(public name : string, public value : number) {}
}
