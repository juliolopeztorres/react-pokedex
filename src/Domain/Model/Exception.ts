
export default class Exception {

  public static create(message: string, context?: object): Exception {
    return new Exception(message, context)
  }

  private constructor(
    public message: string,
    public context?: object
  ) {
  }
}
