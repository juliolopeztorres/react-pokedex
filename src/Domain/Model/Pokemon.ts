export default class Pokemon {

  constructor(
    public id : number,
    public name : string
  ) {}
}
