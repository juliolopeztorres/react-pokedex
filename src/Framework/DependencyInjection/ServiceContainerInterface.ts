
export type ServiceType = 'GetPokemonsUseCase' | 'PokemonRepository' | 'LocalStorage' | string

export default interface ServiceContainerInterface {
  getService(name: ServiceType): any;
}
