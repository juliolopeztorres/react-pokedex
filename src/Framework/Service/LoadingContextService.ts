import { createContext } from 'react'

const LoadingContextService = createContext<[boolean, (status: boolean) => void]>([false, () => {}])

export default LoadingContextService
