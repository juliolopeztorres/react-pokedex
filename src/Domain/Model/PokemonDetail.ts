import BaseStat from "./BaseStat";

export default class PokemonDetail {

  constructor(
    public id : number,
    public name : string,
    public image : string,
    public types : string[],
    public backSprites : string[],
    public baseStats : BaseStat[]
  ) {}
}
