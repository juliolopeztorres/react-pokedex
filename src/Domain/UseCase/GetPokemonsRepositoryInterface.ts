import Pokemon from "../Model/Pokemon";
import PokemonDetail from "../Model/PokemonDetail";
import { Generation } from "../Model/GenerationInfo";

export default interface GetPokemonsRepositoryInterface {
  getByGeneration(generation: Generation): Promise<Pokemon[]>

  getById(id: number): Promise<PokemonDetail>
}
