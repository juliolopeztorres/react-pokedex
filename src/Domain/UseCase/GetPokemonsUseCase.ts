import Exception from "../Model/Exception";
import GetPokemonsRepositoryInterface from "./GetPokemonsRepositoryInterface";
import Pokemon from "../Model/Pokemon";
import PokemonDetail from "../Model/PokemonDetail";
import { Generation } from "../Model/GenerationInfo";

export default class GetPokemonsUseCase {
  constructor(private repository: GetPokemonsRepositoryInterface) {}

  public getAll(generation: Generation): Promise<Pokemon[]>
  {
    return this.repository.getByGeneration(generation)
      .catch((error) => Promise.reject(Exception.create(`Could not recover pokemons for generation ${generation}`, error)))
  }

  public getById(id: number): Promise<PokemonDetail>
  {
    return this.repository.getById(id)
      .catch((error) => Promise.reject(Exception.create(`Could not recover pokemon detail for id ${id}`, error)))
  }
}
