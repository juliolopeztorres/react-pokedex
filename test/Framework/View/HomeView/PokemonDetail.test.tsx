import React from 'react'
import { createRoot, Root } from "react-dom/client"
import { act } from "@testing-library/react";

import Pokemon from "../../../../src/Domain/Model/Pokemon";
import PokemonModelDetail from "../../../../src/Domain/Model/PokemonDetail";
import ServiceContainerInterface, {
  ServiceType
} from "../../../../src/Framework/DependencyInjection/ServiceContainerInterface";
import GetPokemonsUseCase from "../../../../src/Domain/UseCase/GetPokemonsUseCase";
import GetPokemonsRepositoryInterface from "../../../../src/Domain/UseCase/GetPokemonsRepositoryInterface";
import ServiceContainerContextService from "../../../../src/Framework/Service/ServiceContainerContextService";
import LoadingContextService from "../../../../src/Framework/Service/LoadingContextService";
import BaseStat from "../../../../src/Domain/Model/BaseStat";
import Exception from "../../../../src/Domain/Model/Exception";

const logFunctionMock = jest.fn()
jest.mock('../../../../src/Domain/Util/logWithLevel', () => {
  return {
    __esModule: true,
    Level: jest.requireActual('../../../../src/Domain/Util/logWithLevel').Level,
    default: logFunctionMock,
  }
})

import PokemonDetail, { PokemonDetailCallback } from "../../../../src/Framework/View/HomeView/PokemonDetail";

let root: Root
let container: HTMLElement
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);

  root = createRoot(container)
});

afterEach(() => {
  // cleanup on exiting
  act(() => root.unmount())

  container.remove();
});

it('should render', async () => {
  const pokemonDetail = new PokemonModelDetail(
    1,
    'bulbasaur',
    'bulbasaur-image',
    ['plant'],
    ['sprite-1', 'sprite-2'],
    [new BaseStat('Attack', 10)]
  )

  const setIsLoading = jest.fn()

  const onPokemonArtClicked = jest.fn()
  const onCloseClicked = jest.fn()

  const callback: PokemonDetailCallback = {
    onPokemonArtClicked,
    onCloseClicked
  }

  await act( async () => {
    root.render(
      <ServiceContainerContextService.Provider value={new class implements ServiceContainerInterface {
        getService(name : ServiceType) : any {
           expect(name).toBe('GetPokemonsUseCase')

          return new GetPokemonsUseCase(new class implements GetPokemonsRepositoryInterface {
            getByGeneration(generation : string) : Promise<Pokemon[]> {
              return Promise.reject('Not implemented');
            }

            getById(id : number) : Promise<PokemonModelDetail> {
              expect(id).toBe(1)

              return Promise.resolve(pokemonDetail);
            }

          })
        }
      }}>
        <LoadingContextService.Provider value={[false, setIsLoading]}>
          <PokemonDetail pokemon={new Pokemon(1, 'bulbasaur')} callback={callback}/>
        </LoadingContextService.Provider>
      </ServiceContainerContextService.Provider>
    )
  })

  expect(container).toMatchSnapshot()

  expect(setIsLoading).toHaveBeenCalledTimes(2)
  expect(setIsLoading).toHaveBeenNthCalledWith(1, true)
  expect(setIsLoading).toHaveBeenNthCalledWith(2, false)

  const pokemonImage = container.querySelector('div[data-test-id="pokemonImage"]')
  expect(pokemonImage).toBeTruthy()

  expect(onPokemonArtClicked).not.toHaveBeenCalled()
  act(() => {
    pokemonImage!.dispatchEvent(new MouseEvent('click', {bubbles: true}))
  })

  expect(onPokemonArtClicked).toHaveBeenCalledTimes(1)

  const closeBtn = container.querySelector('span[data-test-id="closeBtn"]')
  expect(closeBtn).toBeTruthy()

  expect(onCloseClicked).not.toHaveBeenCalled()
  act(() => {
    closeBtn!.dispatchEvent(new MouseEvent('click', {bubbles: true}))
  })

  expect(onCloseClicked).toHaveBeenCalledTimes(1)

  expect(container).toMatchSnapshot()
});

it('should log if use case fails', async () => {
  const setIsLoading = jest.fn()

  const onPokemonArtClicked = jest.fn()
  const onCloseClicked = jest.fn()

  const callback: PokemonDetailCallback = {
    onPokemonArtClicked,
    onCloseClicked
  }

  const expectedException = Exception.create('Error coming from Repository')

  await act(async () => {
    root.render(
      <ServiceContainerContextService.Provider value={new class implements ServiceContainerInterface {
        getService(name : ServiceType) : any {
          if (name === 'GetPokemonsUseCase') {
            return new GetPokemonsUseCase(new class implements GetPokemonsRepositoryInterface {
              getByGeneration(generation : string) : Promise<Pokemon[]> {
                throw new Error('Not implemented')
              }

              getById(id : number) : Promise<PokemonModelDetail> {
                return Promise.reject(expectedException);
              }
            })
          }

          throw new Error('Unexpected service requested')
        }
      }}>
        <LoadingContextService.Provider value={[false, setIsLoading]}>
          <PokemonDetail pokemon={new Pokemon(1, 'bulbasaur')} callback={callback}/>
        </LoadingContextService.Provider>
      </ServiceContainerContextService.Provider>
    )
  })

  expect(setIsLoading).toHaveBeenCalledTimes(2)
  expect(setIsLoading).toHaveBeenNthCalledWith(1, true)
  expect(setIsLoading).toHaveBeenNthCalledWith(2, false)

  expect(logFunctionMock).toHaveBeenCalledTimes(1)
  expect(logFunctionMock).toHaveBeenLastCalledWith(
    'ðŸ”´',
    'error getting pokemon from repository by id',
    { indentation: 2, context: Exception.create(
        'Could not recover pokemon detail for id 1',
        expectedException
      )}
  )

  expect(container).toMatchSnapshot()
});
