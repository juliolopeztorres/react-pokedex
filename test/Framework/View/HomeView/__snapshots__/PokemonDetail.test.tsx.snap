// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should log if use case fails 1`] = `<div />`;

exports[`should render 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="flex flex-row mb-4"
    >
      <span
        class="
              w-full text-xl lg:text-2xl text-red-600 dark:text-red-400
              text-center font-mono
              underline decoration-dashed underline-offset-8
              "
      >
        #
        1
         
        Bulbasaur
      </span>
      <span
        class="
                w-5 h-6
                text-orange-600 dark:text-orange-100
                text-center
                rounded bg-orange-100 dark:bg-orange-700
                font-mono
                "
        data-test-id="closeBtn"
        style="cursor: pointer;"
      >
        x
      </span>
    </div>
    <div
      class="self-center mb-4 flex flex-col"
      data-test-id="pokemonImage"
      style="cursor: pointer;"
    >
      <img
        alt="Pokemon image"
        class="w-[10rem] md:w-[15rem] dark:text-gray-200 text-xs text-center"
        src="bulbasaur-image"
      />
      <span
        class="italic text-gray-400 text-xs text-center mt-2"
      >
        Click to show alternative art
      </span>
    </div>
    <div
      class="flex flex-row self-center mb-4"
    >
      <div
        class="
                bg-orange-100 border border-orange-400 dark:bg-orange-700
                w-fit px-2 py-1 mr-1 dark:text-gray-200
                last:mr-0 rounded-md hover:bg-orange-300 dark:hover:bg-orange-500
                "
        style="cursor: pointer;"
      >
        plant
      </div>
    </div>
    <div
      class="self-center flex flex-row mb-4"
    >
      <img
        alt="Pokemon back sprite"
        class="w-[10rem]"
        src="sprite-1"
      />
      <img
        alt="Pokemon back sprite"
        class="w-[10rem]"
        src="sprite-2"
      />
    </div>
    <div
      class="
            w-100 border rounded border-orange-400
            bg-orange-100 dark:bg-orange-700
            text-xs md:text-sm dark:text-gray-200
            "
    >
      <ul
        class="px-5 py-2 columns-2"
      >
        <li
          class="mb-1 last:mb-0"
        >
          <span
            class="text-orange-600 dark:text-red-200"
          >
            Attack
          </span>
          : 
          10
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`should render 2`] = `<div />`;
